include/mqtt/broker/offline_message.hpp

```diff
@@ -96,25 +96,85 @@ class offline_messages {
         catch (packet_id_exhausted_error const& e) {
             MQTT_LOG("mqtt_broker", warning)
                 << MQTT_ADD_VALUE(address, &ep)
-                << e.what();
+                << e.what()
+                << " : send the rest messages later.";
         }
         for (auto const& oflm : messages_) {
             oflm.send(ep);
         }
     }
 
+    void send_by_packet_id_release(endpoint_t& ep) {
+        try {
+            auto& idx = messages_.get<tag_seq>();
+            bool finish = false;
+            while (!idx.empty() && !finish) {
+                idx.modify(
+                    idx.begin(),
+                    [&](auto& e) {
+                        auto qos_value = e.pubopts_.get_qos();
+                        e.send(ep);
+                        // if packet_id is consumed, then finish
+                        if (qos_value == qos::at_least_once ||
+                            qos_value == qos::exactly_once) {
+                            finish = true;
+                        }
+                    }
+                );
+                idx.pop_front();
+            }
+        }
+        catch (packet_id_exhausted_error const& e) {
```

Comment:
This loop will run until exception is thrown (packet id exhausted). In my opinion, this is not "exceptional" behaviour. Also, handling an exception is more expensive / performance intensive operation than handling a return value. 

Maybe have an optional acquire_unique_packet_id (called: acquire_unique_packet_id_no_except)

````C++
 /**
     * @brief Acquire the new unique packet id.
     *        If all packet ids are already in use, then throw packet_id_exhausted_error exception.
     *        After acquiring the packet id, you can call acquired_* functions.
     *        The ownership of packet id is moved to the library.
     *        Or you can call release_packet_id to release it.
     * @return packet id
     */
    std::optional<packet_id_t> acquire_unique_packet_id_no_except() {
        LockGuard<Mutex> lck (store_mtx_);
        if (packet_id_.size() == std::numeric_limits<packet_id_t>::max()) std::optional<packet_id_t>();
        if (packet_id_master_ == std::numeric_limits<packet_id_t>::max()) {
            packet_id_master_ = 1U;
        }
        else {
            ++packet_id_master_;
        }

         ....
````

And have existing acquire_packet_unique_id call this method:
````C++
 /**
     * @brief Acquire the new unique packet id.
     *        If all packet ids are already in use, then throw packet_id_exhausted_error exception.
     *        After acquiring the packet id, you can call acquired_* functions.
     *        The ownership of packet id is moved to the library.
     *        Or you can call release_packet_id to release it.
     * @return packet id
     */
    packet_id_t acquire_unique_packet_id() {
        auto result = acquire_unique_packet_id_no_except();
        if (!result) throw packet_id_exhausted_error();
        return result.value();

````C++

Reply:
done :)